server.go

```diff
@@ -78,8 +78,17 @@ type response struct {
 	writer         Writer            // writer to output the raw DNS bits
 }
 
+// handleRefused returns a HandlerFunc that returns REFUSED for every request it gets.
+func handleRefused(w ResponseWriter, r *Msg) {
+	m := new(Msg)
+	m.SetRcode(r, RcodeRefused)
+	w.WriteMsg(m)
+}
+
 // HandleFailed returns a HandlerFunc that returns SERVFAIL for every request it gets.
+// Deprecated: This function is going away.
 func HandleFailed(w ResponseWriter, r *Msg) {
+	println("dns: HandleFailed: this function is deprecated")
```

Comment:
This would be way too noisey if someone is using this externally, possibly printing on every single received packet.

It’s such a small and non-intrusive function that I don’t really see the point in deprecating it honestly.

Reply:
The potential noise is a good point. But then again it would be very easy to replace usage of this deprecated function.

What direction do you want to go here? We could keep it marked deprecated in the documentation but remove the printed warning. Or we could not deprecate it at all. But at that point should we then export the `handleRefused` function to be consistent?

Imo, these simple helper functions, if both exported, only save external users a couple lines of code. And what about the other return codes, where are their helper functions? For those reasons, and since `HandleFailed` is no longer used internally, I do think it should be deprecated. And if so, how are external users to be made aware of the deprecation if there isn't some noise about it? Just my thoughts.

Reply:
Informing developers that they’re using a deprecated symbol is the job of external tooling like staticcheck, that is precisely why the Go Team standardised on a way to mark symbols and packages deprecated. Being loud here would just be us being obnoxious, especially when this function is liable to be called so frequently. Hell it could conceivably lead to a DOS depending on where stdout was going.

The reason it’s exported is that it’s exported, you’re right in that it doesn’t really make much sense to add a new similar function, but comparability guarantees apply to exported symbols.

Like I said, leaving this function around costs us nothing but a few lines dwarfed by the rest of the library. I don’t care much about it bring deprecated, though I don’t think we gain much from that, but it shouldn’t be noisey.